File: mobile/modules/expo-document-camera/android/src/main/java/expo/documentcamera/DocumentCameraModule.kt
package expo.documentcamera

import android.app.Activity
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.util.Log
import com.google.mlkit.vision.documentscanner.GmsDocumentScanning
import com.google.mlkit.vision.documentscanner.GmsDocumentScannerOptions
import com.google.mlkit.vision.documentscanner.GmsDocumentScanningResult
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.io.FileOutputStream

/**
 * Expo module wrapping Google ML Kit Document Scanner API.
 *
 * Provides auto-crop, perspective correction, and deskew via
 * the platform-native document scanner.
 *
 * Also provides image brightness analysis for low-light detection.
 */
class DocumentCameraModule : Module() {
  private val TAG = "ExpoDocumentCamera"
  private var scannerPromise: Promise? = null

  override fun definition() = ModuleDefinition {
    Name("ExpoDocumentCamera")

    AsyncFunction("scanDocumentAsync") { promise: Promise ->
      val context = appContext.reactContext ?: run {
        promise.reject("ERR_NO_CONTEXT", "React context not available")
        return@AsyncFunction
      }

      if (!isScannerAvailable()) {
        promise.reject(
          "DOC_CAM_NOT_SUPPORTED",
          "ML Kit Document Scanner requires Google Play Services with ML Kit support"
        )
        return@AsyncFunction
      }

      val options = GmsDocumentScannerOptions.Builder()
        .setGalleryImportAllowed(false)
        .setPageLimit(1) // Single page for M2-6
        .setScannerMode(GmsDocumentScannerOptions.SCANNER_MODE_FULL)
        .build()

      scannerPromise = promise

      val activity = appContext.currentActivity ?: run {
        promise.reject("ERR_NO_ACTIVITY", "No current activity")
        return@AsyncFunction
      }

      val scanner = GmsDocumentScanning.getClient(options)
      scanner.getStartScanIntent(activity)
        .addOnSuccessListener { intent ->
          activity.startActivityForResult(intent, REQUEST_CODE_SCAN)
        }
        .addOnFailureListener { e ->
          scannerPromise?.reject("DOC_CAM_ERROR", e.message ?: "Scanner failed to start")
          scannerPromise = null
        }
    }

    AsyncFunction("isAvailable") { ->
      isScannerAvailable()
    }

    AsyncFunction("checkImageBrightnessAsync") { uri: String, promise: Promise ->
      try {
        val context = appContext.reactContext ?: throw Exception("No context")
        val imageUri = Uri.parse(uri)

        // Decode a sampled version for speed
        val options = BitmapFactory.Options().apply {
          inSampleSize = 16 // Downsample 16x for quick brightness check
        }

        val bitmap = if (imageUri.scheme == "file") {
          BitmapFactory.decodeFile(imageUri.path, options)
        } else {
          val inputStream = context.contentResolver.openInputStream(imageUri)
          inputStream?.use { BitmapFactory.decodeStream(it, null, options) }
        }

        if (bitmap == null) {
          promise.reject("ERR_LOAD_IMAGE", "Failed to decode image")
          return@AsyncFunction
        }

        val brightness = computeAverageBrightness(bitmap)
        bitmap.recycle()
        promise.resolve(brightness)
      } catch (e: Exception) {
        promise.reject("ERR_BRIGHTNESS", "Failed to check brightness: ${e.message}")
      }
    }

    OnActivityResult { _, _, _, data ->
      handleScanResult(data)
    }
  }

  private fun isScannerAvailable(): Boolean {
    return try {
      val availability = com.google.android.gms.common.GoogleApiAvailability
        .getInstance()
        .isGooglePlayServicesAvailable(appContext.reactContext)
      availability == com.google.android.gms.common.ConnectionResult.SUCCESS
    } catch (e: Exception) {
      Log.w(TAG, "Google Play Services check failed", e)
      false
    }
  }

  private fun computeAverageBrightness(bitmap: Bitmap): Double {
    val width = bitmap.width
    val height = bitmap.height
    val pixels = IntArray(width * height)
    bitmap.getPixels(pixels, 0, width, 0, 0, width, height)

    var totalBrightness = 0.0
    for (pixel in pixels) {
      val r = (pixel shr 16) and 0xFF
      val g = (pixel shr 8) and 0xFF
      val b = pixel and 0xFF
      // Standard luminance: 0.299R + 0.587G + 0.114B
      totalBrightness += 0.299 * r + 0.587 * g + 0.114 * b
    }

    return totalBrightness / pixels.size
  }

  private fun handleScanResult(data: Intent?) {
    if (data == null) {
      scannerPromise?.resolve(null)
      scannerPromise = null
      return
    }

    val result = GmsDocumentScanningResult.fromActivityResultIntent(data)
    if (result == null) {
      scannerPromise?.resolve(null)
      scannerPromise = null
      return
    }

    val pages = result.pages
    if (pages.isNullOrEmpty()) {
      scannerPromise?.resolve(null)
      scannerPromise = null
      return
    }

    // Take the first page (single-photo mode for M2-6)
    val page = pages[0]
    val imageUri = page.imageUri

    // Copy to app cache directory (never camera roll, never upload)
    try {
      val context = appContext.reactContext ?: throw Exception("No context")
      val inputStream = context.contentResolver.openInputStream(imageUri)
        ?: throw Exception("Cannot open image URI")

      val filename = "scan_${System.currentTimeMillis()}.jpg"
      val cacheFile = File(context.cacheDir, filename)

      FileOutputStream(cacheFile).use { outputStream ->
        inputStream.copyTo(outputStream)
      }

      inputStream.close()

      // Get bitmap for dimensions
      val bitmap = android.provider.MediaStore.Images.Media.getBitmap(context.contentResolver, imageUri)

      scannerPromise?.resolve(mapOf(
        "uri" to Uri.fromFile(cacheFile).toString(),
        "width" to bitmap.width,
        "height" to bitmap.height,
        "timestamp" to System.currentTimeMillis()
      ))
    } catch (e: Exception) {
      scannerPromise?.reject("ERR_FILE_SAVE", "Failed to save scanned image: ${e.message}")
    }

    scannerPromise = null
  }

  companion object {
    private const val REQUEST_CODE_SCAN = 0x43414D // "CAM" in hex
  }
}

File: mobile/modules/expo-document-camera/ios/DocumentCameraModule.swift
import ExpoModulesCore
import VisionKit
import Accelerate

/**
 * Expo module wrapping VNDocumentCameraViewController (iOS 13+).
 *
 * Opens the native document scanner which provides auto-crop,
 * perspective correction, and deskew — no post-processing needed.
 *
 * Also provides image brightness analysis for low-light detection.
 */
public class DocumentCameraModule: Module {
  private var scannerPromise: Promise?
  private var scannerVC: VNDocumentCameraViewController?

  public required init(appContext: AppContext) {
    super.init(appContext: appContext)
  }

  public func definition() -> ModuleDefinition {
    Name("ExpoDocumentCamera")

    AsyncFunction("scanDocumentAsync") { (promise: Promise) in
      guard VNDocumentCameraViewController.isSupported else {
        promise.reject(
          "DOC_CAM_NOT_SUPPORTED",
          "Document camera requires iOS 13+"
        )
        return
      }

      guard let currentViewController = self.appContext?.utilities?.currentViewController() else {
        promise.reject("ERR_MISSING_VC", "No active view controller found")
        return
      }

      self.scannerPromise = promise
      self.scannerVC = VNDocumentCameraViewController()
      self.scannerVC!.delegate = self

      currentViewController.present(self.scannerVC!, animated: true)
    }

    AsyncFunction("isAvailable") { () -> Bool in
      if #available(iOS 13, *) {
        return VNDocumentCameraViewController.isSupported
      }
      return false
    }

    AsyncFunction("checkImageBrightnessAsync") { (uri: String, promise: Promise) in
      guard let url = URL(string: uri) ?? URL(fileURLWithPath: uri) as URL? else {
        promise.reject("ERR_INVALID_URI", "Invalid image URI")
        return
      }

      guard let imageSource = CGImageSourceCreateWithURL(url as CFURL, nil),
            let cgImage = CGImageSourceCreateImageAtIndex(imageSource, 0, nil) else {
        promise.reject("ERR_LOAD_IMAGE", "Failed to load image from URI")
        return
      }

      let brightness = Self.averageBrightness(of: cgImage)
      promise.resolve(brightness)
    }
  }

  /// Compute average brightness (0–255) of a CGImage by sampling pixels.
  private static func averageBrightness(of cgImage: CGImage) -> Double {
    let width = cgImage.width
    let height = cgImage.height

    // Downsample to a small thumbnail for speed
    let thumbSize = 64
    let thumbRect = CGRect(x: 0, y: 0, width: thumbSize, height: thumbSize)

    guard let context = CGContext(
      data: nil,
      width: thumbSize,
      height: thumbSize,
      bitsPerComponent: 8,
      bytesPerRow: thumbSize * 4,
      space: CGColorSpaceCreateDeviceRGB(),
      bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
    ) else { return 150 }

    context.draw(cgImage, in: thumbRect)

    guard let pixelData = context.data else { return 150 }
    let data = pixelData.bindMemory(to: UInt8.self, capacity: thumbSize * thumbSize * 4)

    var totalBrightness: Double = 0
    let pixelCount = thumbSize * thumbSize

    for i in 0..<pixelCount {
      let offset = i * 4
      let r = Double(data[offset])
      let g = Double(data[offset + 1])
      let b = Double(data[offset + 2])
      // Standard luminance: 0.299R + 0.587G + 0.114B
      totalBrightness += 0.299 * r + 0.587 * g + 0.114 * b
    }

    return totalBrightness / Double(pixelCount)
  }
}

// MARK: - VNDocumentCameraViewControllerDelegate

extension DocumentCameraModule: VNDocumentCameraViewControllerDelegate {
  public func documentCameraViewController(
    _ controller: VNDocumentCameraViewController,
    didFinishWith scan: VNDocumentCameraScan
  ) {
    scannerVC?.dismiss(animated: true)

    guard scan.pageCount > 0 else {
      // No pages scanned — treat as cancellation
      scannerPromise?.resolve(nil)
      return
    }

    // Take the first page scan (single-photo mode for M2-6)
    let image = scan.imageOfPage(at: 0)

    // Save to app sandbox (cache directory) — never camera roll
    let filename = "scan_\(Int(Date().timeIntervalSince1970 * 1000)).jpg"
    guard let cacheDir = FileManager.default.urls(
      for: .cachesDirectory, in: .userDomainMask
    ).first else {
      scannerPromise?.reject("ERR_FILE_SAVE", "Cannot access cache directory")
      return
    }

    let fileURL = cacheDir.appendingPathComponent(filename)

    guard let data = image.jpegData(compressionQuality: 0.8) else {
      scannerPromise?.reject("ERR_JPEG", "Failed to compress scanned image")
      return
    }

    do {
      try data.write(to: fileURL)
      scannerPromise?.resolve([
        "uri": fileURL.absoluteString,
        "width": image.size.width,
        "height": image.size.height,
        "timestamp": Int(Date().timeIntervalSince1970 * 1000)
      ])
    } catch {
      scannerPromise?.reject("ERR_FILE_SAVE", "Failed to save scanned image: \(error.localizedDescription)")
    }
  }

  public func documentCameraViewControllerDidCancel(
    _ controller: VNDocumentCameraViewController
  ) {
    scannerVC?.dismiss(animated: true)
    // User cancelled — return nil, not an error
    scannerPromise?.resolve(nil)
  }

  public func documentCameraViewController(
    _ controller: VNDocumentCameraViewController,
    didFailWithError error: Error
  ) {
    scannerVC?.dismiss(animated: true)
    scannerPromise?.reject("DOC_CAM_ERROR", error.localizedDescription)
  }
}
